Adds GCE integration test

joaoandremartins opened the pull request:
Uses cloud-init to get the container up and running.

joaoandremartins commented:
There is still an issue with this PR that I'm looking into. For some reason, when the docker run command is ran from the script, I'm reading this from `journalctl -xe`:

Jun 13 18:39:22 openjdk-gce-integration-test docker[946]: Pulling repository gcr.io/sodium-gateway-790/openjdk-gce-integration-test
Jun 13 18:39:22 openjdk-gce-integration-test docker[650]: time="2017-06-13T18:39:22.888382616Z" level=error msg="Not continuing with pull after error: Error: Status 405 trying to pull repository sodium-gateway-790/openjdk-gce-integration-test: \"v1 Registry API is disabled. See https://cloud.google.com/container-registry/docs/support/deprecation-notices\""
Jun 13 18:39:22 openjdk-gce-integration-test docker[946]: /usr/bin/docker: Error: Status 405 trying to pull repository sodium-gateway-790/openjdk-gce-integration-test: "v1 Registry API is disabled. See https://cloud.google.com/container-registry/docs/support/deprecation-notices".
Jun 13 18:39:22 openjdk-gce-integration-test docker[946]: See '/usr/bin/docker run --help'.

Running the command manually works, for some reason.

Also, the instance isn't being cleaned up afterwards.

cassand commented:
I think that you will have to fetch the instance IP and run the runtime common integration tests on it.

Documentation:
[Runtime Common - Integration Tests](https://github.com/GoogleCloudPlatform/runtimes-common/tree/master/integration_tests)

Example:
``` bash
echo "Running integration tests on application that is deployed at $DEPLOYED_APP_URL"
gcloud container builds submit \
  --config ${dir}/integration_test.yaml \
  --substitutions "_DEPLOYED_APP_URL=http://$DEPLOYED_APP_URL" \
  ${dir}
```

joaoandremartins commented:
For completion, here are more log lines (from `sudo journalctl -ex`) from when the container can't run at startup:

**Jun 13 19:23:01 openjdk-gce-integration-test docker[650]: time="2017-06-13T19:23:01.760981481Z" level=warning msg="Error getting v2 registry: Get https://gcr.io/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)"
Jun 13 19:23:01 openjdk-gce-integration-test docker[650]: time="2017-06-13T19:23:01.761471809Z" level=error msg="Attempting next endpoint for pull after error: Get https://gcr.io/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)"**
Jun 13 19:23:04 openjdk-gce-integration-test device_policy_manager[738]: E0613 19:23:04.336135     738 main.go:387] SetInstanceStatus returned status 403 (PERMISSION_DENIED): Request had insufficient authentication scopes.
Jun 13 19:23:04 openjdk-gce-integration-test device_policy_manager[738]: I0613 19:23:04.336513     738 main.go:366] Using InstanceConfig:
Jun 13 19:23:04 openjdk-gce-integration-test device_policy_manager[738]: I0613 19:23:04.337001     738 manager_impl.go:238] Updating device policy. Instance config changed from {metrics_enabled:false target_version_prefix:"" update_scatter_seconds:33 reboot_after_update:false } to {metrics_enabled:false target_version_prefix:"" update_scatter_seconds:0 reboot_after_update:false }
Jun 13 19:23:04 openjdk-gce-integration-test update_engine[737]: [0613/192304:INFO:real_device_policy_provider.cc(70)] Reloading and re-scheduling device policy due to signal received.
**Jun 13 19:23:04 openjdk-gce-integration-test docker[936]: Pulling repository gcr.io/sodium-gateway-790/openjdk-gce-integration-test
Jun 13 19:23:04 openjdk-gce-integration-test docker[650]: time="2017-06-13T19:23:04.851949600Z" level=error msg="Not continuing with pull after error: Error: Status 405 trying to pull repository sodium-gateway-790/openjdk-gce-integration-test: \"v1 Registry API is disabled. See https://cloud.google.com/container-registry/docs/support/deprecation-notices\""
Jun 13 19:23:04 openjdk-gce-integration-test systemd[1]: testapp.service: Main process exited, code=exited, status=125/n/a
Jun 13 19:23:04 openjdk-gce-integration-test docker[936]: /usr/bin/docker: Error: Status 405 trying to pull repository sodium-gateway-790/openjdk-gce-integration-test: "v1 Registry API is disabled. See https://cloud.google.com/container-registry/docs/support/deprecation-notices".
Jun 13 19:23:04 openjdk-gce-integration-test docker[936]: See '/usr/bin/docker run --help'.**

I think the non-bold lines happen as a consequence of the failure to reach out to the v2 registry.

I found [some answers](https://github.com/moby/moby/issues/22635) on the Internet that suggest this could be either due to the MTU or bad DNS server, but none of these appear to be the problem here.

When I try to start the service by myself, either through `sudo systemctl start testapp.service` or using the `/usr/bin/docker run --rm -u 2000 --name=testapp gcr.io/sodium-gateway-790/openjdk-gce-integration-test:latest /bin/sleep 3600` command directly, it works.

aslo commented on scripts/gce_integration_test.sh:
```diff
@@ -0,0 +1,67 @@
+export STAGING_IMAGE=${imageUnderTest}
+envsubst < "Dockerfile.in" > "Dockerfile"
+export TESTED_IMAGE=${imageUrl}
+envsubst < "openjdk-spring-boot.yaml.in" > "openjdk-spring-boot.yaml"
```
This is a kubernetes config file - I don't think it's necessary to do this for the  GCE test.

aslo commented on scripts/gce_integration_test.sh:
```diff
@@ -0,0 +1,67 @@
+pushd ${deployDir}
+export STAGING_IMAGE=${imageUnderTest}
+envsubst < "Dockerfile.in" > "Dockerfile"
+export TESTED_IMAGE=${imageUrl}
```
Could you change these variable names to make it clearer the difference between STAGING_IMAGE and TESTED_IMAGE. I understand that the name STAGING_IMAGE is required for other reasons, but what about STAGING_IMAGE and APPLICATION_IMAGE?

meltsufin commented:
It looks like you need to enable the "Registry API" in the project to be able to pull the test application image.

meltsufin commented on scripts/gce_integration_test_metadata.in:
```diff
@@ -0,0 +1,24 @@
+#cloud-config
```
This file should move to `test-applicaiton/src/main/gce/` following the pattern for kubernetes test.

meltsufin commented on scripts/gce_integration_test_metadata.in:
```diff
@@ -0,0 +1,24 @@
+#cloud-config
+
+users:
+- name: testapp
```
We might have an issue running jetty under this user.

meltsufin commented on scripts/gce_integration_test_metadata.in:
```diff
@@ -0,0 +1,24 @@
+    [Service]
+    Environment="HOME=/home/testapp"
+    ExecStartPre=/usr/share/google/dockercfg_update.sh
+    ExecStart=/usr/bin/docker run --rm -u 2000 --name=testapp ${TESTED_IMAGE}:latest /bin/sleep 3600
```
You need to expose the HTTP port here. Something like `-p 80:8080`. Also, you'll want to remove `/bin/sleep 3600` and rely on the default `CMD`.

aslo commented on scripts/gce_integration_test.sh:
```diff
@@ -0,0 +1,67 @@
+envsubst < "openjdk-spring-boot.yaml.in" > "openjdk-spring-boot.yaml"
+
+echo "Deploying image to Google Container Registry..."
+gcloud docker -- build -t "$imageName" .
```
I think you don't need to use `gcloud docker` for the build and tag steps. You can just do `docker build ...` and `docker tag`.

Also, you can do both of these in a single step by just passing `"$imageUrl"` to the tag option of docker build:
```
docker build -t  "$imageUrl"
```

cassand replied:
The gcloud docker command is necessary in case you use a base image which is located on a private Container Registry (which is the default behavior of build.sh).

aslo replied:
Yep, but it's only necessary when running `gcloud docker -- push`. The other docker actions can be run using vanilla docker

cassand replied:
I think docker build too, to access the base image.

aslo replied:
Ah. Good point!

joaoandremartins commented:
I think I addressed all comments, except for @cassand's one about integration test bootstrap, which I'm delaying until I can figure out why https://gcr.io/v2/ is unreachable by cloud-config, using the configured new user.

joaoandremartins commented:
@meltsufin @aslo @cassand with the latest changes, I think all comments are addressed.
The integration tests are failing with timeout, I suspect because of the enforced firewall rules. Where can I refer to passing integration tests?

joaoandremartins commented on scripts/gce_integration_test.sh:
```diff
@@ -0,0 +1,81 @@
+gcloud docker -- tag "$imageName" "$imageUrl"
+gcloud docker -- push $imageUrl
+
+export CONTAINER_USER_NAME="joeuser"
```
A testuser user is already created, I can rename to something else.

joaoandremartins commented:
In the projects used currently for integration testing, or your own test projects, are the firewall rules being enforced? Or did you have to ask for any exemption?

joaoandremartins commented:
Filed internal issue 63087569 to do something about the firewall rules on my test project.

joaoandremartins commented:
The created instance is now reachable from the Internet but, for some reason, the Container Builder still can't reach it. Looking into this.

joaoandremartins commented:
Once we apply the same kind of GCE enforcer whitelist for the project running the integration test instances, I think we're good to merge this PR.
Although apparently there are still some cases where the instance is unreachable on some IP ranges...

meltsufin commented:
@joaoandremartins Is this good to merge now?

balopat commented:
I had a look at this, two problems:
1.) the branch got unfortunately pretty outdated, the structure of the test-app changed, so it's not working anymore
2.) I'd finish first #148 to remove duplication - this gce_integration script is repeating a lot of the same code what is already heavily duplicated.
